unstructured/ingest/v2/processes/connectors/sharepoint.py

```diff
@@ -0,0 +1,406 @@
+        rel_path = fullpath.replace(self.index_config.path, "")
+        while rel_path[0] == "/":
+            rel_path = rel_path[1:]
+        return FileData(
+            identifier=file.unique_id,
+            connector_type=CONNECTOR_TYPE,
+            source_identifiers=SourceIdentifiers(
+                filename=file.name,
+                fullpath=fullpath,
+                rel_path=rel_path,
+            ),
+            metadata=DataSourceMetadata(
+                url=absolute_url,
+                version=f"{file.major_version}.{file.minor_version}",
+                date_modified=str(date_modified_dt.timestamp()) if date_modified_dt else None,
+                date_created=str(date_created_at.timestamp()) if date_created_at else None,
+                date_processed=str(time()),
+                record_locator={"server_path": file.serverRelativeUrl, "site_url": client.base_url},
+            ),
+            additional_metadata=additional_metadata,
+        )
+
+    def get_root(self, client: "ClientContext") -> "Folder":
+        if path := self.index_config.path:
+            return client.web.get_folder_by_server_relative_path(path)
+        default_document_library = client.web.default_document_library()
+        root_folder = default_document_library.root_folder
+        root_folder = root_folder.get().execute_query()
+        self.index_config.path = root_folder.name
+        return root_folder
+
+    def get_site_url(self, client: "ClientContext") -> str:
+        res = client.web.get().execute_query()
+        return res.url
+
+    def get_site(self, permissions_client: "GraphClient", site_url) -> "Site":
+        return permissions_client.sites.get_by_url(url=site_url).execute_query()
+
+    def get_permissions_items(self, site: "Site") -> list["DriveItem"]:
+        # TODO find a way to narrow this search down by name of drive
+        items: list["DriveItem"] = []
+        drives: list["Drive"] = site.drives.get_all().execute_query()
+        for drive in drives:
+            items.extend(drive.root.children.get_all().execute_query())
+        return items
+
+    def map_permission(self, permission: "Permission") -> dict:
+        return {
+            "id": permission.id,
+            "roles": list(permission.roles),
+            "share_id": permission.share_id,
+            "has_password": permission.has_password,
+            "link": permission.link.to_json(),
+            "granted_to_identities": permission.granted_to_identities.to_json(),
+            "granted_to": permission.granted_to.to_json(),
+            "granted_to_v2": permission.granted_to_v2.to_json(),
+            "granted_to_identities_v2": permission.granted_to_identities_v2.to_json(),
+            "invitation": permission.invitation.to_json(),
+        }
+
+    def enrich_permissions_on_files(self, all_file_data: list[FileData], site_url: str) -> None:
+        logger.debug("Enriching permissions on files")
```

Comment:
Suggesting to move after line 278 because even empty permission_client s produce this log, where permissions won't get ingested. Alternatively, could add a log as "failed due to an error while obtaining permissions_client"